src/resolve/resolved-mdns-browse-services.c

```diff
@@ -0,0 +1,454 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later
+ * Copyright (c) 2022 Koninklijke Philips N.V. */
+
+#include <net/if.h>
+#include "resolved-varlink.h"
+#include "resolved-mdns-browse-services.h"
+#include "resolved-dns-cache.h"
+
+
+int mdns_add_new_service(mDnsServiceSubscriber *ss,  DnsResourceRecord *rr, int owner_family) {
+        _cleanup_(mdns_service_unrefp) mDnsServices *s = NULL;
+
+        assert(ss);
+        assert(rr);
+
+        s = new(mDnsServices, 1);
+
+        if (!s)
+                return log_oom();
+
+        *s = (mDnsServices) {
+                .n_ref = 1,
+                .rr = dns_resource_record_copy(rr),
+                .family = owner_family,
+        };
+
+        LIST_PREPEND(mdns_services, ss->mdns_services, s);
+
+        TAKE_PTR(s);
+
+        return 0;
+}
+
+void mdns_remove_service(mDnsServiceSubscriber *ss, mDnsServices *service) {
+        assert(ss);
+        assert(service);
+
+        LIST_REMOVE(mdns_services, ss->mdns_services, service);
+        service = mdns_service_free(service);
+}
+
+mDnsServices *mdns_service_free(mDnsServices *service) {
+
+        if (!service)
+                return NULL;
+
+        dns_resource_record_unref(service->rr);
+        return mfree(service);
+}
+
+DEFINE_TRIVIAL_REF_UNREF_FUNC(mDnsServices, mdns_service, mdns_service_free);
+
+bool mdns_service_contains(mDnsServices *services, DnsResourceRecord *rr, int owner_family) {
+
+        LIST_FOREACH(mdns_services, service, services) {
+                if (dns_resource_record_equal(rr, service->rr) && service->family == owner_family) {
+                        return true;
+                }
+        }
+        return false;
+}
+
+void mdns_manage_services_answer(mDnsServiceSubscriber *ss, DnsAnswer *answer, int owner_family) {
+        DnsResourceRecord *i = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *array = NULL;
+        _cleanup_(json_variant_unrefp) JsonVariant *entry = NULL;
+        _cleanup_free_ char *name = NULL;
+        _cleanup_free_ char *type = NULL;
+        _cleanup_free_ char *domain = NULL;
+        int r = 0;
```

Comment:
Unnecessary initialization. Please do like the following at the end of the function:
```c
                r = json_variant_append_array(&array, entry);
                if (r < 0)
                        goto finish;
        }

        return 0;

finish:
        log_error_errno(r, "Failed to ...");
        return varlink_error_errno(...);
}
```